synapse/federation/federation_server.py

```diff
@@ -882,32 +932,40 @@ async def _process_incoming_pdus_in_room_inner(
         room_id: str,
         room_version: RoomVersion,
         lock: Lock,
-        latest_origin: str,
-        latest_event: EventBase,
+        latest_origin: Optional[str] = None,
+        latest_event: Optional[EventBase] = None,
     ) -> None:
         """Process events in the staging area for the given room.
 
         The latest_origin and latest_event args are the latest origin and event
-        received.
+        received (or None to simply pull the next event from the database).
         """
 
         # The common path is for the event we just received be the only event in
         # the room, so instead of pulling the event out of the DB and parsing
         # the event we just pull out the next event ID and check if that matches.
-        next_origin, next_event_id = await self.store.get_next_staged_event_id_for_room(
-            room_id
-        )
-        if next_origin == latest_origin and next_event_id == latest_event.event_id:
-            origin = latest_origin
-            event = latest_event
-        else:
+        found_origin = None
+        found_event = None
+        if latest_event is not None and latest_origin is not None:
+            (
+                next_origin,
+                next_event_id,
+            ) = await self.store.get_next_staged_event_id_for_room(room_id)
+            if next_origin == latest_origin and next_event_id == latest_event.event_id:
+                found_origin = latest_origin
+                found_event = latest_event
+
+        if found_origin is None or found_event is None:
             next = await self.store.get_next_staged_event_for_room(
                 room_id, room_version
             )
             if not next:
                 return
 
             origin, event = next
+        else:
+            origin = found_origin
+            event = found_event
```

Comment:
I would simplify this by just clearing latest_event if it is found unacceptable:

```suggestion
        if latest_event is not None and latest_origin is not None:
            (
                next_origin,
                next_event_id,
            ) = await self.store.get_next_staged_event_id_for_room(room_id)
            if next_origin != latest_origin or next_event_id != latest_event.event_id:
                latest_origin = None

        if latest_origin is None or latest_event is None:
            next = await self.store.get_next_staged_event_for_room(
                room_id, room_version
            )
            if not next:
                return

            origin, event = next
        else:
            origin = latest_origin
            event = latest_event
```
ymmv though.